File: solutions/go/03/knapsack/main.go
package main

import "fmt"

func knapsack(weights, values []int, cap int) int {
	dp := make([]int, cap+1)
	for i := 0; i < len(weights); i++ {
		w, v := weights[i], values[i]
		for c := cap; c >= w; c-- {
			if dp[c-w]+v > dp[c] {
				dp[c] = dp[c-w] + v
			}
		}
	}
	return dp[cap]
}

func main() {
	fmt.Println(knapsack([]int{2, 3, 4}, []int{3, 4, 5}, 5))
}

File: solutions/python/03/knapsack/main.py
def knapsack(weights: list, values: list, cap: int) -> int:
    dp = [0] * (cap + 1)
    for w, v in zip(weights, values):
        for c in range(cap, w - 1, -1):
            dp[c] = max(dp[c], dp[c - w] + v)
    return dp[cap]


def main() -> None:
    weights = [2, 3, 4]
    values = [3, 4, 5]
    print(knapsack(weights, values, 5))


if __name__ == "__main__":
    main()
